lib/logger/suppression.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package logger
+
+import (
+	"fmt"
+	"hash/fnv"
+	"sync"
+	"time"
+)
+
+// The suppressingLogger passes logging through to the next logger, unless
+// it's a repeated message or the rate thresholds are exceeded in which case
+// previously seen messages are filtered.
+type suppressingLogger struct {
+	next lowlevelLogger
+
+	mut            sync.Mutex
+	buckets        []suppressingLoggerBucket
+	bucketInterval time.Duration // how often to rotate
+	threshold      int           // threshold for suppression in average messages/bucket
+	active         bool          // whether suppression is currently active
+	lastHash       uint64        // hash of last message
+	lastRepeated   int           // how many times the last message has been repeated
+	stop           chan struct{}
+}
+
+func newSuppressingLogger(next lowlevelLogger, numBuckets, threshold int, bucketInterval time.Duration) *suppressingLogger {
+	l := &suppressingLogger{
+		next:           next,
+		buckets:        make([]suppressingLoggerBucket, numBuckets),
+		bucketInterval: bucketInterval,
+		threshold:      threshold,
+		stop:           make(chan struct{}),
+	}
+	l.buckets[0].reset() // prep for first use
+	return l
+}
+
+func (l *suppressingLogger) Serve() {
+	// Rotate buckets every bucketInterval, also setting/clearing the active
+	// flag.
+
+	t := time.NewTicker(l.bucketInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			l.rotateBuckets()
+		case <-l.stop:
+			return
+		}
+	}
+}
+
+func (l *suppressingLogger) Stop() {
+	close(l.stop)
+}
+
+func (l *suppressingLogger) SetFlags(flags int) {
+	l.next.SetFlags(flags)
+}
+
+func (l *suppressingLogger) SetPrefix(prefix string) {
+	l.next.SetPrefix(prefix)
+}
+
+func (l *suppressingLogger) Output(level int, message string) error {
+	l.mut.Lock()
+	defer l.mut.Unlock()
+
+	hash := messageHash(message)
+	l.buckets[0].seen[hash] = struct{}{}
+
+	if l.suppressSameAsLast(hash) {
+		// Message is the same as the last message
+		return nil
```

Comment:
Does this address any concrete case? I couldn't think of any. And in general I'd rather keep repeat messages if suppression is not active/volume is low. I believe that we should not hide them, but correct them where they occur when someone notices (e.g. the repeat warnings in quic recently).

TLDR: I wouldn't suppress anything unless active/high volume.